### PR TITLE
feat: add automation scheduler timers

### PR DESCRIPTION
**Pull request by Astro-Han**
## Summary

Adds the PR3 backend scheduler slice for issue #950: a per-instance in-memory automation scheduler service for one-shot and interval-after-completion definitions, scheduled unattended execution, and no-overlap skip recording.

The scheduler now owns one interruptible Effect-backed task per scheduled automation, tracks the unattended runs it starts so shutdown can cancel them, and applies recurring stop rules that are safe for this in-memory slice. Native timers are contained behind the scheduler clock/sleep adapter instead of being the service architecture.

## Why

PR1 froze the automation contract and PR2 made manual `runNow` execute safely. This PR adds background firing with the smallest useful runtime behavior while keeping the scheduler shape aligned with the broader supervised backend service design. Persistence, cron, durable locks, worktree placement, frontend, and tool-unhide remain separate later slices.

## Related Issue

Closes part of #950: PR3 scheduler timers + ownership + unattended policy.

## Human Review Status

Pending

## Review Focus

Please focus on the scheduler ownership boundary, interruptible wait/run lifecycle, long-delay waiting, completed one-shot rescheduling, recurring no-overlap behavior, recurring stop=count handling, condition-stop non-scheduling until an evaluator exists, and recurring interval anchoring after either scheduled or manual run completion.

## Risk Notes

The scheduler is intentionally in-memory and per app instance for this slice. It is now modeled as a supervised scheduler service with cancelable per-definition tasks and cancelable scheduler-owned unattended runs, while the actual waiting remains process-local. Durable cross-window locking, persistence, cron, missed-run scan, restart reconciliation, and durable `nextFireAt`/`nextFires` state are deferred to later slices per the frozen plan.

For this PR, `nextFireAt` / `nextFires` remain contract fields and are not a source of truth for the in-memory scheduler queue. The real scheduler ownership lives inside the per-instance scheduler.

Condition-based recurring stops are accepted by the frozen contract, but PR3 does not include a condition evaluator. To avoid turning an unevaluated condition into an infinite unattended loop, condition-stop definitions are not scheduled by this in-memory scheduler until that evaluator slice exists.

Skipped checklist items:
- Visible UI check: no visible UI or copy changed.
- Platform/packaging check: no platform, packaging, updater, signing, shell, or native path behavior changed.
- Docs/release/dependency check: no docs, release notes, dependencies, credentials, generated content, or local file behavior changed.

## How To Verify

```text
Whitespace check: git diff --check -> passed
Automation scheduler regression: bun test test/server/automation-scheduler.test.ts --timeout 30000 -> 26 passed
Automation regression bundle: bun test test/server/automation-runner.test.ts test/server/automation-scheduler.test.ts test/server/automation-routes.test.ts test/server/automation-event-fixtures.test.ts test/server/event-replay.test.ts test/tool/automate.test.ts --timeout 30000 -> 104 passed
Typecheck: bun --cwd packages/opencode typecheck -> passed
```

## Screenshots or Recordings

Not applicable. No visible UI changes.

## Checklist

> **How to use this checklist:**
>
> - Tick a box by replacing `[ ]` with `[x]`. Do not edit, add, or remove items.
> - The bot-applied label items can only be honestly ticked AFTER the PR is opened and the labeler / priority-triage bots have run — return to the PR description and tick them then.
> - Most items are required. The few that are conditional are explicitly marked **(conditional)**; for those, leave unticked if they truly do not apply and explain why in Risk Notes. All other items must be ticked before requesting human review.

- [x] **Type label** — this PR carries exactly one of `bug`, `enhancement`, `task`, `documentation`. Type labels are author-added; the labeler bot does NOT assign them. Add the label in the GitHub UI, then tick this.
- [x] **Routing labels** — this PR carries at least one of `app`, `ui`, `platform`, `harness`, `ci`. The labeler bot assigns these on PR open based on changed paths. Confirm the bot's choice (or override if wrong), then tick this.
- [x] **Priority label** — this PR carries exactly one of `P0`, `P1`, `P2`, `P3`. The priority-triage bot suggests one on PR open. Confirm or override, then tick this.
- [x] Human Review Status above is set to `Pending`, `Approved by @<reviewer>`, or `Not required: <reason>` (default is `Pending`; "not required" is restricted to bot-authored low-risk PRs).
- [x] I linked the related issue, or stated in Summary why there is no issue.
- [x] I described the review focus and any meaningful risks.
- [x] I replaced the example block in How To Verify with the real verification steps and the key result for each.
- [x] I did not introduce unrelated refactors, dependencies, generated files, or file changes beyond the stated scope.
- [ ] **(conditional)** I manually checked visible UI or copy changes when needed, with screenshots or recordings. Leave unticked only if no visible UI or copy changed.
- [ ] **(conditional)** I considered macOS and Windows impact for platform, packaging, updater, signing, paths, shell, or permissions changes. Leave unticked only if no platform/packaging surface was touched.
- [ ] **(conditional)** I called out docs, release notes, dependencies, permissions, credentials, deletion behavior, generated content, or local file changes when relevant. Leave unticked only if none of those surfaces was touched.
- [x] I reviewed the final diff for unrelated changes and suspicious dependency changes.
- [x] I am targeting `dev`, and my PR title and commit messages use Conventional Commits in English.

## Summary by CodeRabbit

* **New Features**
  * Added a scheduler for automations: recurring and one-shot scheduling, missed-schedule handling, and lazy startup.
  * New run-query helpers and ability to record stopped runs with custom timestamps.

* **Bug Fixes**
  * More robust active-run tracking and safer cleanup to avoid stopping newer executions.
  * Better coordination between manual and scheduled runs to prevent overlap.

* **Tests**
  * Comprehensive scheduler and route tests covering scheduling, cancellation, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Astro-Han/pawwork/pull/984?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/opencode/src/automation/scheduler.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+    const fire = (automationID: string, triggeredAt: number) => {
+      timers.delete(automationID)
+      const definition = getOptional(automationID)
+      if (!definition || definition.paused) return
+      if (Automation.hasActiveRun(automationID)) {
+        const stopped = Automation.recordStoppedRun(automationID, "previous_run_awaiting_input", { now: triggeredAt })
+        void Automation.publishRunUpdated(stopped)
+        const next = nextFireAt(definition, triggeredAt)
+        if (definition.kind === "recurring" && next !== null) schedule(definition, next)
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/opencode/src/automation/scheduler.ts]: Fixed in the current head: recurring follow-up scheduling now anchors from `clock.now()` after overlap skips and executor completion, so a delayed fire cannot cascade from a stale `triggeredAt`.

File: packages/opencode/src/automation/scheduler.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    const reschedule = (definition: Automation.Definition) => {
+      cancel(definition.id)
+      if (definition.paused) return
+      const next = nextFireAt(definition, clock.now())
+      if (next !== null) schedule(definition, next)
+    }
+
+    return {
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/opencode/src/automation/scheduler.ts:268]: Fixed in the current head: scheduler construction scans `Automation.list()` and reschedules existing definitions before returning the service.

File: packages/opencode/src/automation/scheduler.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+  export const layer = (options?: Options) => Layer.effect(Service, Effect.sync(() => Service.of(make(options))))
+  export const defaultLayer = layer()
+
+  export function nextFireAt(definition: Automation.Definition, from: number): number | null {
+    if (definition.paused) return null
+    if (definition.kind === "oneshot") return definition.fireAt
+    if (definition.rhythm.kind !== "interval") return null
+    return from + definition.rhythm.everyMs
+  }
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/opencode/src/automation/scheduler.ts]: Fixed in the current head: one-shot `computeNextFireAt` now returns `null` once a run has already triggered at or after the definition fire time, and the regression covers update plus pause/resume after completion.

File: packages/opencode/src/project/instance.ts [hunk 4]
```diff
@@ -123,6 +127,7 @@ export const Instance = {
   },
   async dispose(input?: { mode?: "maintenance" | "force"; onCompleted?: () => void | Promise<void> }) {
     const ctx = Instance.current
+    if ((input?.mode ?? "maintenance") === "maintenance") await (await scheduler()).stopCurrentOwnedRuns()
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/opencode/src/project/instance.ts]: Fixed in 65bf291: scheduler-owned run pre-stop during maintenance disposal is now best-effort across dispose, disposeDirectory, and disposeAll; failures are logged and disposal continues. Added coverage for a throwing pre-stop helper.
